File: src/fib.js
/*************************************************************
Fibonacci

**************************************************************/

//without memoization, this approach was loosely bounded by O(2^n)
//with memoization below, we reduce it to linear O(n)
function fib(num) {
    if (num <= 1) {
        return 1;
    } else {
        return fib(num-1) + fib(num-2);
    }
}

//fib(8);

//top down dynamic programming approach
//function decorator to memoize
function memoize(fun) {
    var cache = {};
    
    //function wrapper checks if input
    //has been cached or not
    function helper(x) {
        //either grab result from cache
        //or compute it
        if (cache[x] === undefined) {
            cache[x] = fun(x);
        }
        //return result that has been 
        //computed
        return cache[x];
    }
    return helper;
}

//wrap original function by returning memoize
var fib = memoize(fib);

fib(8);

/*************************************************************
Recursive

**************************************************************/
nthFibonacci = function(n) {  
    if (n === 0 || n === 1) {
        return n;
    }
    return nthFibonacci(n-1) + nthFibonacci(n-2);
};

/*************************************************************
Iterative

**************************************************************/

nthFibonacci = function(n) {  
  var x = 0;
  var y = 1;
  var sum = 1;
  
  for (var i = 2; i <= n; i++) {
    sum = x + y;
    x = y;
    y = sum;
  }
  
  return (n === 0 ? 0 : sum);
};

//alt iterative solution 

nthFibonacci = function(n) {  
  var fib = [0, 1];
  for (var i = 2; i < n + 1; i++) {
    fib[i] = fib[i-1] + fib[i-2];
  }
  return fib[n];
};

/*************************************************************
Other versions

**************************************************************/

var fibRecurs = function(n){
  return n < 2 ? n : fibRecurs(n-1) + fibRecurs(n-2);
};
​
var fibMem = function(n){
  for(var i = 0; i <= n; i++){
    fibMem.mem = i < 2 ? i : fibMem.mem[i-2] + fibMem.mem[i-1];
  }
  return fibMem.mem[n];
};
​
fibMem.mem = [];
​
var fibSpace = function(n){
  var mem = [0,1];
  for(; n > 1; n--){
    mem.push(mem.shift() +mem[0]);
  }
  return mem[n];
};

File: ruby/fib.rb

# classic

def fib(num)
  if num <= 1
    return 1
  else
    return fib(num-1) + fib(num-2)
  end
end

## Another way

def fib(n)
  return n if n < 2
 
  fib(n-1) + fib(n-2)
end

# iterative

def nFib(n)
  x, y, sum = 0, 1, 1
  for i in 2..n
    sum = x + y
    x = y
    y = sum
  end
  n === 0 ? 0 : sum
end

def nthFib(n)
  fib = [0,1]
  for i in 2..n
    fib[i] = fib[i-1] + fib[i-2]
  end
  fib[n]
end

### More Recursive
# without memoization, this approach was loosely bounded by O(2^n)
# with memoization below, we reduce it to linear O(n)
def fib(n)
  n < 2 ? n : fib(n-1) + fib(n-2)
end

def memoize(fn)
  cache = {}
  
  # Since a class inherits directly from its singleton class, 
  # this is where we want to add class methods dynamically 
  # when metaprogramming.
  fxn = singleton_class.instance_method(fn)
  
  define_singleton_method fn do |*args|
    unless cache.include?(args)
      cache[args] = fxn.bind(self).call(*args)
    end
    cache[args]
  end
end

memoize :fib
fib(8)

### Simpler one with memoization
@cache = [0,1]

def fib(n)
  return @cache[n] if @cache[n]
 
  @cache[n] = fib(n-1) + fib(n-2)
end

### another iterative one
def fibIter(n)
  memo = []
 
  (0..n).each do |i|
    memo[i] = i < 2 ? i : memo[i-1] + memo[i-2]
  end
  memo[n]
end















